### ISSUE TITLE
(Angular) Add schematics for StateAdapt

### ISSUE THREAD

**Issue by pBouillon** [state-adapt/state-adapt#66]
Hi,

I am currently in the process of migrating older Angular apps to v17, and I find the usage of schematics really helpful for libraries that support it. This made me wonder if StateAdapt could also have some as well.

For example, when executing:

```bash
~$ ng add @state-adapt/angular
```

The schematic could automatically install `@state-adapt/angular`, `@state-adapt/core` and `@state-adapt/rxjs`, and update the providers array to provide `defaultStoreProvider`. If relevant, it could also initialize the devtools configuration, maybe configurable based on an additionnal argument passed to `ng add`:

```diff
import { ApplicationConfig } from '@angular/core';

+ import { defaultStoreProvider } from '@state-adapt/angular';

export const appConfig: ApplicationConfig = {
  providers: [
+    defaultStoreProvider,
  ],
};
```

Or even:

```diff
import { ApplicationConfig } from '@angular/core';

+ import { actionSanitizer, stateSanitizer } from '@state-adapt/core';
+ import { provideStore } from '@state-adapt/angular';

export const appConfig: ApplicationConfig = {
  providers: [
+    provideStore({
+      devtools: (window as any)?.__REDUX_DEVTOOLS_EXTENSION__?.({
+        actionSanitizer,
+        stateSanitizer,
+      }),
+      showSelectors: false,
+    }),
  ],
};
```

```diff
{
  "name": "my-app",
  "version": "0.0.0",
  "scripts": {
    ...
  },
  "private": true,
  "dependencies": {
    "@angular/animations": "^17.0.3",
    "@angular/common": "^17.0.3",
    "@angular/compiler": "^17.0.3",
    "@angular/core": "^17.0.3",
    "@angular/forms": "^17.0.3",
    "@angular/platform-browser": "^17.0.3",
    "@angular/platform-browser-dynamic": "^17.0.3",
    "@angular/router": "^17.0.3",
+  "@state-adapt/angular": "^2.0.0",
+  "@state-adapt/core": "^2.0.0",
+  "@state-adapt/rxjs": "^2.0.0",
    "rxjs": "~7.8.1",
    "tslib": "^2.6.2",
    "zone.js": "~0.14.0"
  },
  "devDependencies": {
    ...
  }
}
```

This feature could introduce a better support to the Angular package, allowing for easier migrations or installations of StateAdapt in their project.

> Side question: the repository and package is `state-adapt` but the docs is `StateAdapt`, what's the correct way of calling the library?

**Comment by mfp22** [comment]
I've never been great at this. I'll have to ask ChatGPT how to get started with this :)
I enjoy the AST work, just not sure how to bundle it all together in something that can operate on the whole codebase

**Comment by pBouillon** [comment]
I don't know either, but I would gladly help you on this if you want!

**Comment by mfp22** [comment]
Believe it or not, I merged your code for this without testing it myself :) It looked better than nothing, and I will test it before releasing.

The docs will need to be updated as well. If you feel like doing that, great, if not, I'll do it after I verify that this works.

**Comment by pBouillon** [comment]
Well, I hope it works correctly then!

I'll gladly submit another PR for the docs. Would you prefer to assess that the schematic is working appropriately before the submission?

**Comment by mfp22** [comment]
That would be great. I can test before publishing the new docs version.

**Comment by mfp22** [comment]
You can create a PR for docs before I test. I have to run `npm run publish:docs` so the PR won't make a public change.

**Comment by pBouillon** [comment]
PR submitted! Let's hope everything works fine now

While working on it, I noticed that it's not that easy to get the project up and running. Maybe a `CONTRIBUTING.md` guide and/or a [devcontainer](https://containers.dev/) could ease the process.

It has a pretty good integration with GitHub code spaces (for example, this is [the one used by NgRx](https://github.com/ngrx/platform/tree/main/.devcontainer)).

I also noticed that Angular is used in version 14, with modules mostly. Since the v17 has already been released, it could be great to catch up with the updated version as well.

All of this has nothing to do with the addition of a schematic but that's just some ideas I had while working on this and I wanted to share!

**Comment by mfp22** [comment]
Yeah I need to run nx migrations so I can add signals. I'll do that after the entity bug. Then we'll see if it got easier to set up. I probably need to add `.npmrc` because I think the version matters a lot. Did you set up on Windows, or do you think a devcontainer would be a good idea regardless?

Thanks for the contributions. I wonder if it's time for me to stop committing directly to `main` 🤔. Probably when I get 2 simultaneous PRs, because then it could start to be hard to keep track of. For now I am enjoying the lack of friction for myself :)

**Comment by pBouillon** [comment]
An `.npmrc` is a great start!

> Did you set up on Windows

To be honest, since it was just a bit of markdown, I did it without running the projects and investigating their dependencies and requirements

> do you think a devcontainer would be a good idea regardless?

This format is supported by VS Code and GitHub, it can be a good way to ease contributions while also making the dependencies a bit more explicit for local setup

That's just my opinion, other alternatives might be suitable too!

About contributions, you're most welcome for your reactivity and follow-up on those! If any issue is still open or if there is room for improvement elsewhere (examples, documentation, feature, etc.) I would gladly help on those as well if that's an option for you (moreover, the documentation won't interfere with your work, the commits on `main` are still possible!)

**Comment by mfp22** [comment]
An installation schematic is no longer required because it's ready to use upon `npm install`. 

There is now an evmrc too.

Schematics will be added for future upgrade migrations.
